File: demo_dagger/src/main/java/com/mandria/android/mvp/example/di/components/AppComponent.java
package com.mandria.android.mvp.example.di.components;

import com.mandria.android.mvp.di.CoreModule;
import com.mandria.android.mvp.example.MVPApplication;
import com.mandria.android.mvp.example.activities.MainActivity;
import com.mandria.android.mvp.example.di.modules.AppModule;
import com.mandria.android.mvp.example.di.modules.PresenterModule;

import javax.inject.Singleton;

import dagger.BindsInstance;
import dagger.Component;


@Singleton
@Component(modules = {
        CoreModule.class,
        AppModule.class,
        PresenterModule.class
})
public interface AppComponent {

    void inject(MainActivity mainActivity);

    @Component.Builder
    interface Builder {

        /**
         * Add the application object to the graph.
         */
        @BindsInstance
        Builder application(MVPApplication application);

        /**
         * Build the component.
         */
        AppComponent build();
    }
}

File: demo_dagger_kotlin/src/main/kotlin/com/mandria/android/mvp/kotlin/example/di/AppComponent.kt
package com.mandria.android.mvp.kotlin.example.di

import android.app.Application
import com.mandria.android.mvp.di.CoreModule
import com.mandria.android.mvp.kotlin.example.activities.MainActivity
import dagger.BindsInstance
import dagger.Component
import javax.inject.Singleton

@Singleton
@Component(
    modules = [AppModule::class, CoreModule::class, PresenterModule::class]
)
interface AppComponent {

    fun inject(mainActivity: MainActivity)

    /** Component builder which only needs the application objects as dependency for the modules. */
    @Component.Builder
    interface Builder {

        /** Add the application object to the graph. */
        @BindsInstance
        fun application(application: Application): Builder

        /** Build the component. */
        fun build(): AppComponent
    }
}